File: internal/orchestrator/heartbeat_test.go
package orchestrator

import (
	"log/slog"
	"os"
	"testing"
	"time"

	"github.com/Z-M-Huang/openhive/internal/domain"
	"github.com/Z-M-Huang/openhive/internal/ws"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func newTestMonitor(t *testing.T, checkInterval, unhealthyTimeout time.Duration) *HeartbeatMonitorImpl {
	t.Helper()
	logger := slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelWarn}))
	return NewHeartbeatMonitorWithIntervals(nil, logger, checkInterval, unhealthyTimeout)
}

func TestProcessHeartbeat_StoresStatusCorrectly(t *testing.T) {
	m := newTestMonitor(t, time.Second, 5*time.Second)

	agents := []domain.AgentHeartbeatStatus{
		{AID: "aid-001", Status: domain.AgentStatusIdle},
	}
	m.ProcessHeartbeat("team-001", agents)

	status, err := m.GetStatus("team-001")
	require.NoError(t, err)
	assert.Equal(t, "team-001", status.TeamID)
	assert.True(t, status.IsHealthy)
	assert.Equal(t, 1, len(status.Agents))
	assert.Equal(t, "aid-001", status.Agents[0].AID)
	assert.WithinDuration(t, time.Now(), status.LastSeen, time.Second)
}

func TestGetStatus_ReturnsNotFoundForUnknownTeam(t *testing.T) {
	m := newTestMonitor(t, time.Second, 5*time.Second)

	_, err := m.GetStatus("nonexistent-team")
	require.Error(t, err)
	var nfe *domain.NotFoundError
	assert.ErrorAs(t, err, &nfe)
}

func TestGetAllStatuses_ReturnsAllTeams(t *testing.T) {
	m := newTestMonitor(t, time.Second, 5*time.Second)

	m.ProcessHeartbeat("team-alpha", []domain.AgentHeartbeatStatus{})
	m.ProcessHeartbeat("team-beta", []domain.AgentHeartbeatStatus{})
	m.ProcessHeartbeat("team-gamma", []domain.AgentHeartbeatStatus{})

	all := m.GetAllStatuses()
	assert.Len(t, all, 3)
	assert.Contains(t, all, "team-alpha")
	assert.Contains(t, all, "team-beta")
	assert.Contains(t, all, "team-gamma")
}

func TestUnhealthyDetection_TriggersCallbackAfterTimeout(t *testing.T) {
	// Use very short intervals for testing
	m := newTestMonitor(t, 10*time.Millisecond, 20*time.Millisecond)

	unhealthyCalled := make(chan string, 1)
	m.SetOnUnhealthy(func(teamID string) {
		unhealthyCalled <- teamID
	})

	// Inject a stale status to trigger the unhealthy check
	m.statuses.Store("team-stale", &domain.HeartbeatStatus{
		TeamID:    "team-stale",
		Agents:    nil,
		LastSeen:  time.Now().Add(-100 * time.Millisecond),
		IsHealthy: true,
	})

	m.StartMonitoring()
	defer m.StopMonitoring()

	select {
	case teamID := <-unhealthyCalled:
		assert.Equal(t, "team-stale", teamID)
	case <-time.After(500 * time.Millisecond):
		t.Fatal("unhealthy callback was not called within timeout")
	}
}

func TestUnhealthyDetection_MarksStatusAsUnhealthy(t *testing.T) {
	m := newTestMonitor(t, 10*time.Millisecond, 20*time.Millisecond)
	m.SetOnUnhealthy(func(teamID string) {})

	// Inject an old heartbeat
	m.statuses.Store("team-old", &domain.HeartbeatStatus{
		TeamID:    "team-old",
		LastSeen:  time.Now().Add(-200 * time.Millisecond),
		IsHealthy: true,
	})

	m.StartMonitoring()
	defer m.StopMonitoring()

	// Wait for monitor to run
	time.Sleep(100 * time.Millisecond)

	status, err := m.GetStatus("team-old")
	require.NoError(t, err)
	assert.False(t, status.IsHealthy)
}

func TestHealthyAfterHeartbeat_ResetsUnhealthyTimer(t *testing.T) {
	m := newTestMonitor(t, 10*time.Millisecond, 20*time.Millisecond)

	unhealthyCount := 0
	m.SetOnUnhealthy(func(teamID string) {
		unhealthyCount++
	})

	// Start with stale status
	m.statuses.Store("team-recover", &domain.HeartbeatStatus{
		TeamID:    "team-recover",
		LastSeen:  time.Now().Add(-200 * time.Millisecond),
		IsHealthy: true,
	})

	m.StartMonitoring()
	defer m.StopMonitoring()

	time.Sleep(50 * time.Millisecond)

	// Send a fresh heartbeat — this should mark the team healthy again
	m.ProcessHeartbeat("team-recover", []domain.AgentHeartbeatStatus{})

	status, err := m.GetStatus("team-recover")
	require.NoError(t, err)
	// After fresh heartbeat, should be healthy again
	assert.True(t, status.IsHealthy)
}

func TestStopMonitoring_StopsTickerCleanly(t *testing.T) {
	m := newTestMonitor(t, 10*time.Millisecond, 50*time.Millisecond)
	m.StartMonitoring()
	time.Sleep(20 * time.Millisecond)
	m.StopMonitoring()

	// Calling stop multiple times should not panic
	m.StopMonitoring()
}

func TestUnhealthyDetection_PublishesEvent(t *testing.T) {
	published := make(chan domain.Event, 5)
	mockBus := &mockEventBus{publishFn: func(e domain.Event) { published <- e }}

	logger := slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelWarn}))
	m := NewHeartbeatMonitorWithIntervals(mockBus, logger, 10*time.Millisecond, 20*time.Millisecond)

	// Inject stale status
	m.statuses.Store("team-pub", &domain.HeartbeatStatus{
		TeamID:    "team-pub",
		LastSeen:  time.Now().Add(-200 * time.Millisecond),
		IsHealthy: true,
	})

	m.StartMonitoring()
	defer m.StopMonitoring()

	var gotStateChange bool
	deadline := time.NewTimer(500 * time.Millisecond)
	defer deadline.Stop()
outer:
	for {
		select {
		case evt := <-published:
			if evt.Type == domain.EventTypeContainerStateChanged {
				gotStateChange = true
				break outer
			}
		case <-deadline.C:
			break outer
		}
	}
	assert.True(t, gotStateChange, "EventTypeContainerStateChanged should be published")
}

func TestConvertAgentStatuses(t *testing.T) {
	wsAgents := []ws.AgentStatus{
		{AID: "aid-001", Status: "idle", Detail: "waiting", ElapsedSeconds: 1.5, MemoryMB: 128.0},
		{AID: "aid-002", Status: "running", Detail: "processing", ElapsedSeconds: 5.0, MemoryMB: 256.0},
	}

	result := ConvertAgentStatuses(wsAgents)
	require.Len(t, result, 2)
	assert.Equal(t, "aid-001", result[0].AID)
	assert.Equal(t, domain.AgentStatusIdle, result[0].Status)
	assert.Equal(t, 1.5, result[0].ElapsedSeconds)
	assert.Equal(t, 128.0, result[0].MemoryMB)
	assert.Equal(t, "aid-002", result[1].AID)
}

// --- Mock EventBus ---

type mockEventBus struct {
	publishFn func(domain.Event)
}

func (m *mockEventBus) Publish(e domain.Event) {
	if m.publishFn != nil {
		m.publishFn(e)
	}
}
func (m *mockEventBus) Subscribe(eventType domain.EventType, handler func(domain.Event)) string {
	return ""
}
func (m *mockEventBus) FilteredSubscribe(eventType domain.EventType, filter func(domain.Event) bool, handler func(domain.Event)) string {
	return ""
}
func (m *mockEventBus) Unsubscribe(id string) {}
func (m *mockEventBus) Close()                {}

File: backend/src/orchestrator/heartbeat.test.ts
/**
 * Tests for HeartbeatMonitorImpl and convertAgentStatuses.
 *
 * Uses fake timers to control setInterval/setTimeout behaviour.
 * All tests use newHeartbeatMonitorWithIntervals (jitter disabled) so
 * startMonitoring() creates the interval immediately.
 *
 * NOTE: We use vi.advanceTimersByTime() (synchronous) — NOT
 * vi.advanceTimersByTimeAsync() which does not work with bun's vitest.
 * After advancing fake time we flush pending microtasks with repeated
 * `await Promise.resolve()` to ensure async callbacks have run.
 */

import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import {
  HeartbeatMonitorImpl,
  newHeartbeatMonitor,
  newHeartbeatMonitorWithIntervals,
  convertAgentStatuses,
} from './heartbeat.js';
import type { HeartbeatLogger } from './heartbeat.js';
import type { EventBus } from '../domain/interfaces.js';
import type { Event } from '../domain/types.js';
import { NotFoundError } from '../domain/errors.js';
import type { AgentStatus } from '../ws/messages.js';
import type { AgentHeartbeatStatus } from '../domain/types.js';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Flushes microtasks by yielding to the event loop several times. */
async function flushMicrotasks(): Promise<void> {
  for (let i = 0; i < 10; i++) {
    await Promise.resolve();
  }
}

/** Creates a silent no-op logger for tests. */
function makeLogger(): HeartbeatLogger {
  return {
    debug: () => undefined,
    warn: () => undefined,
  };
}

/** Creates a mock EventBus that records published events. */
function makeEventBus(): EventBus & { published: Event[] } {
  const published: Event[] = [];
  return {
    published,
    publish(event: Event): void {
      published.push(event);
    },
    subscribe: () => 'sub-id',
    filteredSubscribe: () => 'sub-id',
    unsubscribe: () => undefined,
    close: () => undefined,
  };
}

/** Creates sample agent heartbeat statuses. */
function makeAgents(count: number = 2): AgentHeartbeatStatus[] {
  return Array.from({ length: count }, (_, i) => ({
    aid: `aid-00${i + 1}`,
    status: 'idle' as const,
    detail: '',
    elapsed_seconds: 0,
    memory_mb: 100,
  }));
}

// ---------------------------------------------------------------------------
// processHeartbeat — stores status and marks healthy
// ---------------------------------------------------------------------------

describe('processHeartbeat', () => {
  it('stores status and marks healthy', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);
    const agents = makeAgents(2);

    monitor.processHeartbeat('tid-001', agents);

    const status = monitor.getStatus('tid-001');
    expect(status.team_id).toBe('tid-001');
    expect(status.is_healthy).toBe(true);
    expect(status.agents).toHaveLength(2);
    expect(status.last_seen).toBeInstanceOf(Date);
    // last_seen should be close to now
    expect(Date.now() - status.last_seen.getTime()).toBeLessThan(1_000);
  });

  it('overwrites a previous status on repeated heartbeats', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);

    monitor.processHeartbeat('tid-001', makeAgents(1));
    const firstSeen = monitor.getStatus('tid-001').last_seen;

    monitor.processHeartbeat('tid-001', makeAgents(3));

    const status = monitor.getStatus('tid-001');
    expect(status.agents).toHaveLength(3);
    expect(status.is_healthy).toBe(true);
    // last_seen should be >= first call (same millisecond or later)
    expect(status.last_seen.getTime()).toBeGreaterThanOrEqual(firstSeen.getTime());
  });

  it('publishes heartbeat_received event', () => {
    const bus = makeEventBus();
    const monitor = newHeartbeatMonitorWithIntervals(bus, makeLogger(), 30_000, 90_000);
    const agents = makeAgents(1);

    monitor.processHeartbeat('tid-002', agents);

    expect(bus.published).toHaveLength(1);
    const event = bus.published[0];
    expect(event).toBeDefined();
    expect(event!.type).toBe('heartbeat_received');
    if (event!.payload.kind === 'heartbeat_received') {
      expect(event!.payload.team_id).toBe('tid-002');
      expect(event!.payload.status.is_healthy).toBe(true);
    }
  });

  it('does not publish event when eventBus is null', () => {
    // Just verify no error is thrown when eventBus is null.
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);
    expect(() => monitor.processHeartbeat('tid-003', makeAgents(1))).not.toThrow();
  });
});

// ---------------------------------------------------------------------------
// getStatus — returns stored status or throws NotFoundError
// ---------------------------------------------------------------------------

describe('getStatus', () => {
  it('returns stored status', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);
    monitor.processHeartbeat('tid-010', makeAgents(2));

    const status = monitor.getStatus('tid-010');
    expect(status.team_id).toBe('tid-010');
  });

  it('throws NotFoundError for unknown team', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);
    expect(() => monitor.getStatus('tid-unknown')).toThrow(NotFoundError);
  });
});

// ---------------------------------------------------------------------------
// getAllStatuses — returns snapshot of all statuses
// ---------------------------------------------------------------------------

describe('getAllStatuses', () => {
  it('returns empty object when no heartbeats received', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);
    expect(monitor.getAllStatuses()).toEqual({});
  });

  it('returns all known statuses', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);
    monitor.processHeartbeat('tid-001', makeAgents(1));
    monitor.processHeartbeat('tid-002', makeAgents(2));

    const all = monitor.getAllStatuses();
    expect(Object.keys(all)).toHaveLength(2);
    expect(all['tid-001']).toBeDefined();
    expect(all['tid-002']).toBeDefined();
  });
});

// ---------------------------------------------------------------------------
// checkHealth — marks stale containers unhealthy
// ---------------------------------------------------------------------------

describe('checkHealth', () => {
  it('marks stale container unhealthy', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);

    monitor.processHeartbeat('tid-stale', makeAgents(1));
    expect(monitor.getStatus('tid-stale').is_healthy).toBe(true);

    monitor.injectStaleStatus('tid-stale');
    monitor.checkHealth();

    expect(monitor.getStatus('tid-stale').is_healthy).toBe(false);
  });

  it('does not mark fresh container unhealthy', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);

    monitor.processHeartbeat('tid-fresh', makeAgents(1));
    monitor.checkHealth();

    expect(monitor.getStatus('tid-fresh').is_healthy).toBe(true);
  });

  it('publishes container_state_changed event on unhealthy transition', () => {
    const bus = makeEventBus();
    const monitor = newHeartbeatMonitorWithIntervals(bus, makeLogger(), 30_000, 90_000);

    monitor.processHeartbeat('tid-stale2', makeAgents(1));
    bus.published.length = 0; // clear the heartbeat_received event

    monitor.injectStaleStatus('tid-stale2');
    monitor.checkHealth();

    const stateEvents = bus.published.filter((e) => e.type === 'container_state_changed');
    expect(stateEvents).toHaveLength(1);
    const event = stateEvents[0];
    expect(event).toBeDefined();
    if (event!.payload.kind === 'container_state_changed') {
      expect(event!.payload.team_id).toBe('tid-stale2');
      expect(event!.payload.state).toBe('failed');
    }
  });
});

// ---------------------------------------------------------------------------
// onUnhealthy callback — fires for unhealthy containers
// ---------------------------------------------------------------------------

describe('onUnhealthy callback', () => {
  it('fires callback when container becomes unhealthy', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);
    const unhealthyTeams: string[] = [];

    monitor.setOnUnhealthy((teamID) => {
      unhealthyTeams.push(teamID);
    });

    monitor.processHeartbeat('tid-cb', makeAgents(1));
    monitor.injectStaleStatus('tid-cb');
    monitor.checkHealth();

    expect(unhealthyTeams).toEqual(['tid-cb']);
  });

  it('does not fire callback for healthy containers', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);
    const unhealthyTeams: string[] = [];

    monitor.setOnUnhealthy((teamID) => {
      unhealthyTeams.push(teamID);
    });

    monitor.processHeartbeat('tid-healthy', makeAgents(1));
    monitor.checkHealth();

    expect(unhealthyTeams).toHaveLength(0);
  });
});

// ---------------------------------------------------------------------------
// Healthy-to-unhealthy transition fires only once
// ---------------------------------------------------------------------------

describe('healthy-to-unhealthy transition fires only once', () => {
  it('callback fires once even if checkHealth is called multiple times', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 30_000, 90_000);
    const calls: string[] = [];

    monitor.setOnUnhealthy((teamID) => calls.push(teamID));

    monitor.processHeartbeat('tid-once', makeAgents(1));
    monitor.injectStaleStatus('tid-once');

    // First check: healthy → unhealthy (callback fires)
    monitor.checkHealth();
    // Second check: already unhealthy (callback must NOT fire again)
    monitor.checkHealth();

    expect(calls).toHaveLength(1);
    expect(calls[0]).toBe('tid-once');
  });
});

// ---------------------------------------------------------------------------
// convertAgentStatuses — maps correctly
// ---------------------------------------------------------------------------

describe('convertAgentStatuses', () => {
  it('maps known status values correctly', () => {
    const wsAgents: AgentStatus[] = [
      { aid: 'aid-001', status: 'idle', elapsed_seconds: 10, memory_mb: 50 },
      { aid: 'aid-002', status: 'busy', detail: 'working', elapsed_seconds: 30, memory_mb: 120 },
      { aid: 'aid-003', status: 'starting', elapsed_seconds: 1, memory_mb: 30 },
    ];

    const result = convertAgentStatuses(wsAgents);

    expect(result).toHaveLength(3);
    expect(result[0]).toEqual({
      aid: 'aid-001',
      status: 'idle',
      detail: '',
      elapsed_seconds: 10,
      memory_mb: 50,
    });
    expect(result[1]).toEqual({
      aid: 'aid-002',
      status: 'busy',
      detail: 'working',
      elapsed_seconds: 30,
      memory_mb: 120,
    });
    expect(result[2]).toEqual({
      aid: 'aid-003',
      status: 'starting',
      detail: '',
      elapsed_seconds: 1,
      memory_mb: 30,
    });
  });

  it('maps unknown status to error', () => {
    const wsAgents: AgentStatus[] = [
      { aid: 'aid-bad', status: 'unknown_value', elapsed_seconds: 0, memory_mb: 0 },
    ];

    const result = convertAgentStatuses(wsAgents);
    expect(result[0]!.status).toBe('error');
  });

  it('handles all valid AgentStatusType values', () => {
    const validStatuses = ['idle', 'busy', 'starting', 'stopped', 'error'] as const;
    for (const s of validStatuses) {
      const wsAgents: AgentStatus[] = [{ aid: 'aid-x', status: s, elapsed_seconds: 0, memory_mb: 0 }];
      const result = convertAgentStatuses(wsAgents);
      expect(result[0]!.status).toBe(s);
    }
  });

  it('handles empty input', () => {
    const result = convertAgentStatuses([]);
    expect(result).toHaveLength(0);
  });
});

// ---------------------------------------------------------------------------
// startMonitoring — begins interval checking
// ---------------------------------------------------------------------------

describe('startMonitoring', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('calls checkHealth after each interval tick', async () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 1_000, 5_000);
    monitor.processHeartbeat('tid-interval', makeAgents(1));

    // Spy on checkHealth to count calls
    const checkHealthSpy = vi.spyOn(monitor, 'checkHealth');

    monitor.startMonitoring();

    // Advance past two intervals
    vi.advanceTimersByTime(2_100);
    await flushMicrotasks();

    expect(checkHealthSpy).toHaveBeenCalledTimes(2);

    monitor.stopMonitoring();
    checkHealthSpy.mockRestore();
  });

  it('is idempotent — multiple calls do not create multiple intervals', async () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 1_000, 5_000);
    const checkHealthSpy = vi.spyOn(monitor, 'checkHealth');

    monitor.startMonitoring();
    monitor.startMonitoring(); // second call must be a no-op
    monitor.startMonitoring(); // third call must be a no-op

    vi.advanceTimersByTime(2_100);
    await flushMicrotasks();

    // Only 2 ticks from one interval, not 6 from three intervals
    expect(checkHealthSpy).toHaveBeenCalledTimes(2);

    monitor.stopMonitoring();
    checkHealthSpy.mockRestore();
  });
});

// ---------------------------------------------------------------------------
// stopMonitoring — clears interval
// ---------------------------------------------------------------------------

describe('stopMonitoring', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('stops checkHealth from being called after stop', async () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 1_000, 5_000);
    const checkHealthSpy = vi.spyOn(monitor, 'checkHealth');

    monitor.startMonitoring();

    // Let two ticks run
    vi.advanceTimersByTime(2_100);
    await flushMicrotasks();
    expect(checkHealthSpy).toHaveBeenCalledTimes(2);

    monitor.stopMonitoring();

    // Advance more time — no further calls expected
    vi.advanceTimersByTime(5_000);
    await flushMicrotasks();
    expect(checkHealthSpy).toHaveBeenCalledTimes(2);

    checkHealthSpy.mockRestore();
  });

  it('is idempotent — multiple stop calls are safe', () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 1_000, 5_000);
    monitor.startMonitoring();
    monitor.stopMonitoring();
    expect(() => monitor.stopMonitoring()).not.toThrow();
    expect(() => monitor.stopMonitoring()).not.toThrow();
  });

  it('startMonitoring after stop is a no-op', async () => {
    const monitor = newHeartbeatMonitorWithIntervals(null, makeLogger(), 1_000, 5_000);
    const checkHealthSpy = vi.spyOn(monitor, 'checkHealth');

    monitor.startMonitoring();
    monitor.stopMonitoring();

    // Try to restart — should not create a new interval
    monitor.startMonitoring();
    vi.advanceTimersByTime(2_100);
    await flushMicrotasks();

    // No calls because both the original interval and the re-start are stopped/no-op
    expect(checkHealthSpy).toHaveBeenCalledTimes(0);

    checkHealthSpy.mockRestore();
  });
});

// ---------------------------------------------------------------------------
// newHeartbeatMonitor — factory function
// ---------------------------------------------------------------------------

describe('newHeartbeatMonitor', () => {
  it('creates a monitor that implements HeartbeatMonitor interface', () => {
    const monitor = newHeartbeatMonitor(null, makeLogger());
    expect(monitor).toBeInstanceOf(HeartbeatMonitorImpl);
    expect(typeof monitor.processHeartbeat).toBe('function');
    expect(typeof monitor.getStatus).toBe('function');
    expect(typeof monitor.getAllStatuses).toBe('function');
    expect(typeof monitor.setOnUnhealthy).toBe('function');
    expect(typeof monitor.startMonitoring).toBe('function');
    expect(typeof monitor.stopMonitoring).toBe('function');
  });
});
